doc/GUIDE.md

```diff
@@ -58,31 +58,72 @@ Besides these options, it is also possible to filter the output of jextract usin
 that jextract has. See the section on [filtering](#filtering) for a more detailed overview. See also the full
 list of command line options [here](#command-line-option-reference).
 
+Most of the code that jextract generates will be available through a single class. By default,
+the name of that class is derived from the name of the main header file that is passed to
+jextract. For example, if the header file is named `mylib.h`, then the derived class name
+will be `mylib_h`. The class name can be set explicitly using the `--header-class-name`
+option as well.
+
+Besides the main header class that is generated, jextract also generates several
+separate files for certain declarations in the C header file. Namely, structs, unions,
+function pointer types, and typedefs of struct or union types result in additional files
+being generated. (All of these are discussed in more detail in the
+[Using The Code Generated By Jextract section](#using-the-code-generated-by-jextract)).
+
 Jextract assumes that the version of a native library that a project uses is relatively stable.
 Therefore, jextract is intended to be run once, and then for the generated sources to be added to the project.
-Jextract only needs to be run again when the native library, or jextract itself are updated. (This is also
-the workflow that jextract itself follows: jextract depends on the [libclang](https://clang.llvm.org/docs/LibClang.html)
-native library in order to parse C sources).
+Jextract only needs to be run again when the native library, or jextract itself are updated.
+(This is also the workflow that jextract itself follows: jextract depends on the
+[libclang](https://clang.llvm.org/docs/LibClang.html) native library in order to parse C sources).
+
+### Preprocessor Definitions
```

Comment:
IMHO both this and the following section is not important enough to appear so early in the document. If we went straight to "Using The Code Generated By Jextract" we would not lose anything - then after that section, we can provide more details about various aspects/options.

Reply:
Not sure... I know what you mean, but knowing how to run the generated code seems fairly important. It seems that that is something someone might want to try out after the first example. Unfortunately, this requires understanding library loading, which is fairly complex.

Also, the comment about feeding jextract the right header file seems pretty important since user may run into errors when running jextract otherwise (we've seen many questions about that, e.g. when someone tries to extract some arbitrary Windows header instead of `Windows.h`).

But also, I realized that this guide may not just be used by first-time readers, but also as a reference by more experienced users, in which case it seems nice to have info about command line flags mostly in one place.

Reply:
While it's true that knowing which header to extract is important, I think having a whole section on "preprocessor definitions" feels like having a very cornery section in a place where in reality you want something more general. E.g. extracting windows.h is tricky, because of order dependencies, but I'm not sure the user needs to be aware of -D here. Just mentioning that headers are complex beasts and can depend on each other in unpredictable ways (and refer to the doc of the library that needs to be extracted) should be good enough guidance at this point.

As for library loading I'm not sure... again, yes, we probably need to say something like "you need to ensure the library you want to call is loaded first" - but I'm not sure we need the full shebang of library loading options. In other places you added references to stuff that is further expanded upon in later sections - e.g. 

```
Besides these options, it is also possible to filter the output of jextract using one of the --include-XXX options that jextract has. See the section on [filtering](https://github.com/JornVernee/jextract/blob/Guide/doc/GUIDE.md#filtering) for a more detailed overview.
```

Seems like something similar is required here.

Reply:
Still not sure. The section on pre-processor definitions is already fairly small, and, in some rare cases, `-D` may be required for extraction to work as well. Do we really want to put ~half this section elsewhere? It seems this information is relevant to 'Running Jextract' as well.

For library loading, a user needs to know what happens to the string they pass to `--library`, which requires understanding the distinction between library names/paths, and the fact that the user needs to set `LD_LIBRARY_PATH`/`DYLD_LIBRARY_PATH`/`PATH`. Even for someone who's experienced with native library loading, that seems like important information?

I think in both cases, we would end up splitting just part of the information into a separate section. Even if not ideal, I think it's better to keep all the info together. So, if we can't get away with moving these sections out-of-line completely, which I don't think we can, I'm inclined to leave things as they are in this area.

Maybe we can improve the situation by have a very basic hello world example up front?

Reply:
IMHO is not about moving "half sections", but about writing a section that is correct for the purpose in which it appears in the document. We seem to have two requirements for users getting started with jextract:
* explain that the header on which it is run matters (this can literally be one sentence)
* explain that libraries must be loaded (I think here we can get away by providing something that works in the 80% use cases, plus some comment on LD_LIBRARY_PATH, and refer everything else to a more specialized section. E.g. I don't think here we should be talking about different lookups, in which orders symbols are looked up etc.).

This way you can have a document that serves both "green" developers and more experienced one (the latter will want to keep going after the section illustrating how to use bindings). At least, IMHO.

Reply:
I personally like the new changes a lot! I find the "running jextract" section very informative, providing all the major info w/o overstaying its welcome. Also it points to several other sections in the doc, so it acts a bit as a mental map, which is good.

I have some reservation about "unsupported features" being down in "Advanced" - should that be the last section in "Using jextract bindings" instead? (e.g. first we show what works, and then if the reader was expecting something else, we have a section for that too). Or, maybe pull it more "up front". Right now it seems kind of buried in other stuff, even thought it seems generally useful.

Reply:
Ok, I'll keep it like this then.

I'll move the unsupported feature to the end of the 'using' section.